WordPress/gutenberg
# Template inspector: Surface related patterns

## What?
Fixes https://github.com/WordPress/gutenberg/issues/44582

This adds a pattern selector in the template inspector panel. Also removes the replace button added in https://github.com/WordPress/gutenberg/pull/54609.

## Why?
This makes it much easier for users to switch out templates and template parts for patterns.

## How?
- Reimplements TemplateList in the template inspector.
- Removes some of the unnecessary filters.
- Splits filters and mapping into different functions.

## Testing Instructions
1. Open the site editor
2. Open a template or a template part
3. Open the inspector
4. Confirm that you can see a list of other suitable templates in the inspector
5. Select a pattern
6. Confirm that the content of the template/template part changes
7. Save the changes
8. Open the frontend of the site and confirm that your changes are saved.

## Screenshots or screencast 
<img width="1512" alt="Screenshot 2023-10-05 at 15 31 33" src="https://github.com/WordPress/gutenberg/assets/275961/6a8dbb64-e8bc-4bad-a4e5-53baf20a24e7">
